reference/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -0,0 +1,332 @@
+        }
+    }
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        const auto x_row = agg.get_const_data()[i];
+        for (size_type j = 0; j < x_dim[1]; j++) {
+            x->at(x_row, j) += b->at(i, j);
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_AMGX_PGM_RESTRICT_APPLY_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void prolong_applyadd(std::shared_ptr<const ReferenceExecutor> exec,
+                      const Array<IndexType> &agg,
+                      const matrix::Dense<ValueType> *b,
+                      matrix::Dense<ValueType> *x)
+{
+    auto x_vals = x->get_values();
+    const auto x_stride = x->get_stride();
+    auto x_dim = x->get_size();
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        const auto b_row = agg.get_const_data()[i];
+        for (size_type j = 0; j < x_dim[1]; j++) {
+            x->at(i, j) += b->at(b_row, j);
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_AMGX_PGM_PROLONGATE_APPLY_KERNEL);
+
+
+template <typename IndexType>
+void match_edge(std::shared_ptr<const ReferenceExecutor> exec,
+                const Array<IndexType> &strongest_neighbor,
+                Array<IndexType> &agg)
+{
+    auto agg_vals = agg.get_data();
+    auto strongest_neighbor_vals = strongest_neighbor.get_const_data();
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        if (agg_vals[i] == -1) {
+            auto neighbor = strongest_neighbor_vals[i];
+            if (neighbor != -1 && strongest_neighbor_vals[neighbor] == i) {
+                agg_vals[i] = i;
+                agg_vals[neighbor] = i;
+                // Use the smaller index as agg point
+            }
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_MATCH_EDGE_KERNEL);
+
+
+template <typename IndexType>
+void count_unagg(std::shared_ptr<const ReferenceExecutor> exec,
+                 const Array<IndexType> &agg, size_type *num_unagg)
+{
+    IndexType unagg = 0;
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        unagg += (agg.get_const_data()[i] == -1);
+    }
+    *num_unagg = unagg;
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_COUNT_UNAGG_KERNEL);
+
+
+template <typename IndexType>
+void renumber(std::shared_ptr<const ReferenceExecutor> exec,
+              Array<IndexType> &agg, size_type *num_agg)
+{
+    const auto num = agg.get_num_elems();
+    Array<IndexType> agg_map(exec, num + 1);
+    auto agg_vals = agg.get_data();
+    auto agg_map_vals = agg_map.get_data();
+    for (size_type i = 0; i < num + 1; i++) {
+        agg_map_vals[i] = 0;
+    }
+    for (size_type i = 0; i < num; i++) {
+        agg_map_vals[agg_vals[i]] = 1;
+    }
+    components::prefix_sum(exec, agg_map_vals, num + 1);
+    for (size_type i = 0; i < num; i++) {
+        agg_vals[i] = agg_map_vals[agg_vals[i]];
+    }
+    *num_agg = agg_map_vals[num];
```

Comment:
Here as well, `num_agg` is `size_type` while `agg_map_vals[num]` is `IndexType`.